### DISCUSSION TITLE
Telemetry Graph not Plotting

### DISCUSSION THREAD

**Discussion by Overloader6** [nasa/openmct#4136] [Community Q&A] [answered]
So, since the old KSP plugin seems to be dead I'm trying to write a new one. I managed to receive telemetry data, and it shows in openmct... Kind of. In the top left of the attached screenshot it shows the current value I get from the telemachus api.
![image](https://user-images.githubusercontent.com/22007229/126382775-af77cb54-25c3-4711-9f9e-1de4feb67cd6.png)
But as you can see, no Graph is being plotted and I don't get any errors in the console. I'll put the code snippets that I believe are relevant below

My telemetry Provider: 
```javascript
provider: {
    supportsSubscribe: function (domainObject) {
        return domainObject.type === 'kerbal.telemetry';
    },
    subscribe: function (domainObject, callback) {
        listener[domainObject.identifier.key] = callback;
        socket.send(JSON.stringify({
            '+': [domainObject.identifier.key]
        }));
        return function unsubscribe() {
            delete listener[domainObject.identifier.key];
            socket.send(JSON.stringify({
                '-': [domainObject.identifier.key]
            }));
        };
    }
}
```

the sockets onmessage:
```javascript
socket.onmessage = function (event) {
    const data = JSON.parse(event.data);
    for (const k in data) {
        if (data.hasOwnProperty(k)) {
            listener[k]({
                timestamp: event.timeStamp,
                value: data[k] * 100,
                id: k
            });
        }
    }
}
```

My dictionary.json
```json
{
  "name": "Kerbal Space Program",
  "key": "rootdir",
  "subdirs": [
    {
      "name": "Telemetry",
      "key": "teldir",
      "measurements": [
        {
          "name": "Throttle",
          "key": "f.throttle",
          "values": [
            {
              "key": "value",
              "name": "Value",
              "units": "%",
              "format": "float",
              "min": 0,
              "max": 100,
              "hints": {
                "range": 1
              }
            },
            {
              "key": "utc",
              "source": "timestamp",
              "name": "Timestamp",
              "format": "utc",
              "hints": {
                "domain": 1
              }
            }
          ]
        }
      ]
    }
  ]
}
```

The composite provider for my telemetry domain Object:
```javascript
const measurementProvider = {
    appliesTo: function (domainObject) {
        return domainObject.identifier.namespace === 'vck.zll' && domainObject.identifier.key === 'teldir';
    },
    load: function (domainObject) {
        return getDictionary().then(function (dictionary) {
            return dictionary.subdirs.filter(function (m) {
                return m.key === domainObject.identifier.key;
            })[0].measurements.map(function (m) {
                return {
                    namespace: 'vck.zll',
                    key: m.key
                };
            });
        });
    }
}
```

The part of the object provider that's responsible for the telemetry domain object:
```javascript
const measurement = dictionary.subdirs.filter(function (m) {
    return m.key === 'teldir';
})[0].measurements.filter(function (m) {
    return m.key === identifier.key;
})[0];
if (measurement !== undefined) {
    return {
        identifier: identifier,
        name: measurement.name,
        type: 'kerbal.telemetry',
        telemetry: {
            values: measurement.values
        },
        location: 'vck.zll:rootdir:teldir'
    };
}
```

**Comment by Overloader6** [comment]
If it helps with reproducing it, here's my gitlab repository https://gitlab.com/Overloader/ksp-openmct/-/tree/dev

**Comment by akhenry** [comment]
@Voreck Code looks good at first glance, nice job! Could you provide an example of the telemetry data that you receive over the websocket? Depending on how the data is structured, the `onMessage` handler may need to work slightly differently.

**Comment by Overloader6** [comment]
@akhenry Honestly kinda surprised that it looks good, pretty much the first time I do anything in JS
So a console.log of the event in onmessage gives me this 

![image](https://user-images.githubusercontent.com/22007229/129980489-3195f053-714e-41e2-8cd2-902635390f03.png)
Screenshot of the log since copy pasting it causes it to loose its formatation, maybe im doing something wrong

**Comment by akhenry** [comment, marked as answer]
OK, cool. So the problem is `Event.timeStamp`. It doesn't look like Telemachus provides timestamps associated with the telemetry values that it emits. You're using `Event.timeStamp` instead, but this cannot be used for this purpose because [it's an offset relative to the creation of the document](https://developer.mozilla.org/en-US/docs/Web/API/Event/timeStamp). JavaScript represents time as [an offset relative to the UNIX epoch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now). So, when your value is interpreted it evaluates to 5855 ms since January 1st 1970. If you were to set the time conductor to January 1st 1970 you would probably see your data.

So! The simple solution is just to use `Date.now()` for the timestamp value, instead of `Event.timestamp`. `Date.now()` will give you a timestamp value relative to the UNIX epoch.

**Comment by Overloader6** [comment]
This works, thanks, I wish I knew it was that easy!
Well, i'll still probably have to search for another way to do this since KSP is a game where you can simply advance time but still, at least now I know what the problem is

**Comment by jvigliotta** [comment]
Hello all, going to convert this to a discussion as this could help out some others as well. Thanks!
